### PR TITLE
DNM: rgw: split rgw-admin

### PR DESCRIPTION
**Pull request by cache-nez**
Signed-off-by: Ekaterina Goltsova cache-nez@yandex.ru

### CODE REVIEW

File: src/rgw/CMakeLists.txt [hunk 1]
```diff
@@ -190,8 +190,16 @@ add_dependencies(radosgw cls_rgw cls_lock cls_refcount
 install(TARGETS radosgw DESTINATION bin)
 
 set(radosgw_admin_srcs
-  rgw_admin.cc
```
Comment (oritwas) [inline src/rgw/CMakeLists.txt]: please use 2 spaces for identation

File: src/rgw/rgw_admin.cc [hunk 2]
```diff
@@ -1061,2264 +67,371 @@ class StoreDestructor {
...
+  boost::optional<std::string> compression_type;
 
-void set_quota_info(RGWQuotaInfo& quota, int opt_cmd, int64_t max_size, int64_t max_objects,
-                    bool have_max_size, bool have_max_objects)
-{
-  switch (opt_cmd) {
-    case OPT_QUOTA_ENABLE:
-    case OPT_GLOBAL_QUOTA_ENABLE:
-      quota.enabled = true;
+  int ret = parse_commandline_parameters(args, user_id, tenant, access_key, subuser, secret_key, user_email,user_op,
```
Comment (oritwas) [inline src/rgw/rgw_admin.cc]: I don't like this, too many parameters.
I suggest starting by parsing the command, than for each command parse the parameter for it.
This way we won't have it in main.

File: src/rgw/rgw_admin.cc [hunk 3]
```diff
@@ -1061,2264 +67,371 @@ class StoreDestructor {
...
+                                         sync_from_rm, sync_from_all, sync_from_all_specified, source_zone_name,
+                                         tier_type, tier_type_specified, tier_config_add, tier_config_rm, index_pool,
+                                         data_pool, data_extra_pool, placement_index_type, index_type_specified,
+                                         compression_type, role_name, path, assume_role_doc, policy_name,
+                                         perm_policy_doc, path_prefix);
+  if (ret != 0)
+    return ret;
 
-      // falling through on purpose
+  ret = parse_command(access_key, gen_access_key, secret_key, gen_secret_key, args, opt_cmd, metadata_key, tenant,
```
Comment (oritwas) [inline src/rgw/rgw_admin.cc]: see my previous comment

File: src/rgw/rgw_admin_opt_role.h [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+
+#include "rgw_rest_conn.h"
+#include "rgw_realm_watcher.h"
+
+#include "common/ceph_json.h"
+#include "rgw_role.h"
+#include "rgw_admin_common.h"
+
+int handle_opt_role_get(const string& role_name, const string& tenant, CephContext *context,
```
Comment (oritwas) [inline src/rgw/rgw_admin_opt_role.h]: I suggest have a base class for admin command than one for each admin command type (role, user, multisite ...)
.you can look how we do it for Rest API see rgw_rest.h rgw_rest_s3.h and rgw_rest_admin.h
I think something like command(enum opt) with method parse, execute , get_error

**Comment by oritwas** [comment]
Hi,
you have a changed Beast submodule, please remove it

**Comment by cache-nez** [comment]
Moving those changes to https://github.com/ceph/ceph/pull/20351 (couldn't sign those commits).

**Pull request closed without merging**